### PR TITLE
4.x: Fix Delay hanging due to wrong disposable-chain management

### PR DESCRIPTION
**Pull request by akarnokd**
This PR fixes a bug in `Delay` that can lead to an unwanted cancellation of the drain loop. In addition, the `IStopwatch.Elapsed` should be consistently read while holding the same lock.

The bug is that the `_cancelable` can be reset by the `IDisposable` of the very first task starting the drain loop while that drain loop already set the `IDisposable` for the next round. The first assignment then cancels the second run's cancellation token which then throws an `ObjectDisposedException` and the loop quits for good. Any queued items and terminal signals will then never be emitted.

The solution is to conditionally set that first `IDisposable` if the target is still null, otherwise the subsequent runs have set a newer one and nothing has to be done. The disposability is still intact in both cases.

To be able to do this, the original subscription containers had to be replaced by fields so that the new atomic `Disposable.TryX` methods can be employed.

In the pre-patch version, running the `DelayTest.Delay_DateTimeOffset_DefaultScheduler` with a loop a million times and with no practical time shift (FromMilliseconds(0)) triggered the bug within the first few hundred rounds. With the patch, the same test succeeded all million rounds. I did not change the test because it would take a minute to complete and also code coverage tends to choke on such high-volume tests in my experience.

Fixes: #382

### CODE REVIEW

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs [hunk 1]
```diff
@@ -74,22 +74,31 @@ public override IDisposable Run(TParent parent)
...
-                    var sourceSubscription = new SingleAssignmentDisposable();
-                    _sourceSubscription = sourceSubscription;
-                    sourceSubscription.Disposable = parent._source.SubscribeSafe(this);
+                    Disposable.SetSingle(ref _sourceSubscription, parent._source.SubscribeSafe(this));
 
-                    return StableCompositeDisposable.Create(_sourceSubscription, _cancelable);
+                    return this;
+                }
+
+                protected override void Dispose(bool disposing)
```
Comment (danielcweber) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs]: Move those two disposables and Dispose into the base class?

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs [hunk 2]
```diff
@@ -521,7 +541,10 @@ public L(Absolute parent, IObserver<TSource> observer, IDisposable cancel)
 
                 protected override void RunCore(Absolute parent)
                 {
-                    _cancelable.Disposable = parent._scheduler.Schedule(parent._dueTime, Start);
+                    // ScheduleDrain might have already set a newer disposable
+                    // using TrySetSerial would cancel it, stopping the emission
+                    // and hang the consumer
+                    Disposable.TrySetSingle(ref _cancelable, parent._scheduler.Schedule(parent._dueTime, Start));
```
Comment (danielcweber) [inline Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs]: _cancelable was as SerialDisposable before, even if this works, consider TrySetSerial throughout to reveal intention more consistently.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs (hunk 1 above)
Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs]: The class is ugly, didn't want to accidentally break it.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs (hunk 2 above)
Comment (akarnokd) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs]: That was the bug: _cancelable was SerialDisposable and SerialDisposable always forcefully overwrites and disposes the previous IDisposable it contains.

Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs]: Ah, alright.

File: Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs (hunk 1 above)
Comment (danielcweber) [reply Rx.NET/Source/src/System.Reactive/Linq/Observable/Delay.cs]: Should be good for now, might be an optimization PR for later.

**Pull request merged**